src/svgdom/util/finder.cpp

```diff
@@ -7,107 +7,189 @@
 using namespace svgdom;
 
 namespace{
-class CacheCreator : virtual public svgdom::const_visitor{
+class cache_creator : virtual public svgdom::const_visitor{
 public:
-	std::map<std::string, finder::element_info> cache;
+	std::unordered_map<std::string, const element*> element_by_id_cache;
+	std::unordered_map<std::string, std::list<const element*>> elements_by_class_name_cache;
+	std::unordered_map<std::string, std::list< const element*>> elements_by_tag_name_cache;
+	std::unordered_map<std::string, style_stack> style_stack_by_id_cache;
 	
 	style_stack styleStack;
-	
-	void addToCache(const svgdom::element& e){
+
+	void add_to_element_by_id_cache(const svgdom::element& e){
+		if(!e.id.empty()){
+			this->element_by_id_cache.insert(std::make_pair(e.id, &e));
+		}
+	}
+
+	void add_to_elements_by_class_name_cache(const svgdom::element& e, const svgdom::styleable& s){
+		for (const auto& class_name : s.classes) {
+			auto it = elements_by_class_name_cache.find(class_name);
+
+			if (it != elements_by_class_name_cache.end()) {
+				it->second.push_back(&e);
+			} else {
+				std::list<const element*> elements = {&e};
+				elements_by_class_name_cache.insert(std::make_pair(class_name, elements));
+			}
+		}
+	}
+
+	void add_to_elements_by_tag_name_cache(const svgdom::element& e, const svgdom::styleable& s){
+		if (!s.get_tag().empty()) {
```

Comment:
actually the `get_tag()` method should probably be also a part of `svgdom::element`. And, thus, we would not need reference to `styleable` here. Because every element has a tag, not only styleable elements. So, we need to introduce the virtual `element::get_tag()`. I have just did it, see latest commit on master branch. Just merge the latest master to your forked repo.